### PR TITLE
[Feature/agent_framework] Add Get Workflow API to retrieve a stored template by workflow id

### PR DESCRIPTION
**Pull request by joshpalis**
### Description
Closes https://github.com/opensearch-project/flow-framework/pull/263 in favor of this PR

Adds a Get Workflow API that retrieves a workflow from the Global Context index. We've decided to instead just rename the Get Workflow State API implementation and keep this and the Get Workflow API separate (separate handlers, separate request response classes, separate transport action handlers). I have updated the `RestGetWorkflowAction` -> `RestGetWorkflowStateAction`, `GetWorkflowTransportAction` -> `GetWorkflowStateTransportAction`, `GetWorkflowRequest` -> `GetWorkflowStateRequest`, etc.

The name `RestGetWorkflowAction`, `GetWorkflowTransportAction` will be used for this PR's API

Creating a workflow : 
```
curl -i -XPOST "localhost:9200/_plugins/_flow_framework/workflow" -H "Content-Type:application/json" --data '{"name":"create-connector-register-deploy-model","description":"test case","use_case":"TEST_CASE","version":{"template":"1.0.0","compatibility":["2.12.0","3.0.0"]},"workflows":{"provision":{"nodes":[{"id":"workflow_step_1","type":"create_connector","user_inputs":{"name":"OpenAI Chat Connector","description":"The connector to public OpenAI model service for GPT 3.5","version":"1","protocol":"http","parameters":{"endpoint":"api.openai.com","model":"gpt-3.5-turbo"},"credential":{"openAI_key":"12345"},"actions":[{"action_type":"predict","method":"POST","url":"https://${parameters.endpoint}/v1/chat/completions"}]}},{"id":"workflow_step_2","type":"register_remote_model","previous_node_inputs":{"workflow_step_1":"connector_id"},"user_inputs":{"name":"openAI-gpt-3.5-turbo","function_name":"remote","description":"test model"}},{"id":"workflow_step_3","type":"deploy_model","previous_node_inputs":{"workflow_step_2":"model_id"}}],"edges":[{"source":"workflow_step_1","dest":"workflow_step_2"},{"source":"workflow_step_2","dest":"workflow_step_3"}]}}}'

HTTP/1.1 201 Created
X-OpenSearch-Version: OpenSearch/3.0.0-SNAPSHOT (opensearch)
content-type: application/json; charset=UTF-8
content-length: 38

{"workflow_id":"__uoRowBLOZlW2X2X4im"}
```

Using the API : 
```
curl -i -XGET "localhost:9200/_plugins/_flow_framework/workflow/__uoRowBLOZlW2X2X4im?pretty"
HTTP/1.1 200 OK
X-OpenSearch-Version: OpenSearch/3.0.0-SNAPSHOT (opensearch)
content-type: application/json; charset=UTF-8
content-length: 2484

{
  "name" : "create-connector-register-deploy-model",
  "description" : "test case",
  "use_case" : "TEST_CASE",
  "version" : {
    "template" : "1.0.0",
    "compatibility" : [
      "2.12.0",
      "3.0.0"
    ]
  },
  "workflows" : {
    "provision" : {
      "user_params" : { },
      "nodes" : [
        {
          "id" : "workflow_step_1",
          "type" : "create_connector",
          "previous_node_inputs" : { },
          "user_inputs" : {
            "version" : "1",
            "parameters" : {
              "endpoint" : "api.openai.com",
              "model" : "gpt-3.5-turbo"
            },
            "credential" : {
              "openAI_key" : "AgV46GVJBK4J44APXADtgnCL3TpjQn59H+D2LBqffMkNOXcAXwABABVhd3MtY3J5cHRvLXB1YmxpYy1rZXkAREF1Q2ova3V5MFFZTEFhNERFMTVnMTFkL294MVhpRThVbnhrUHRNa25La2FFR0plSUhSSGRYZmxyWHJ1UmxnVEFDZz09AAEABkN1c3RvbQAUAAAAgAAAAAz4p8I4uGUKqBgB9/QAMDD0u3curGLdlCnrFHA/DVwS/RglaBi8AR5zZ8brKLReZz0YMoUpz/WJtk6KS0qZJwIAABAABJG/gOTF0lfyyCaxDSa1AX33c/1Kcj4hdExYfrqsod+UsRXA5v3hEIYePQi5ocm1/////wAAAAEAAAAAAAAAAAAAAAEAAAAFe3SLotpwFZUcsTiNmxOrsRUeasZ1AGcwZQIwRYI4qrpwzybOuQ0X3C99Oz612jfVx6m0PJg8iOaezz6LYcXU+AiawjFxQLeDtW5pAjEA5Tj7d7G3ubFg8drykv2VT5u9bATBHP0CVWDi2fgrH6etA65kDY3NIqOcvGiwkT7H"
            },
            "actions" : [
              {
                "method" : "POST",
                "action_type" : "predict",
                "url" : "https://${parameters.endpoint}/v1/chat/completions"
              }
            ],
            "description" : "The connector to public OpenAI model service for GPT 3.5",
            "protocol" : "http",
            "name" : "OpenAI Chat Connector"
          }
        },
        {
          "id" : "workflow_step_2",
          "type" : "register_remote_model",
          "previous_node_inputs" : {
            "workflow_step_1" : "connector_id"
          },
          "user_inputs" : {
            "name" : "openAI-gpt-3.5-turbo",
            "description" : "test model",
            "function_name" : "remote"
          }
        },
        {
          "id" : "workflow_step_3",
          "type" : "deploy_model",
          "previous_node_inputs" : {
            "workflow_step_2" : "model_id"
          },
          "user_inputs" : { }
        }
      ],
      "edges" : [
        {
          "source" : "workflow_step_1",
          "dest" : "workflow_step_2"
        },
        {
          "source" : "workflow_step_2",
          "dest" : "workflow_step_3"
        }
      ]
    }
  }
}
```

### Issues Resolved
https://github.com/opensearch-project/flow-framework/issues/170
Part of #88 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/flowframework/transport/GetWorkflowStateAction.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import org.opensearch.action.ActionType;
+
+import static org.opensearch.flowframework.common.CommonValue.TRANSPORT_ACTION_NAME_PREFIX;
+
+/**
+ * External Action for public facing RestGetWorkflowStateAction
+ */
+public class GetWorkflowStateAction extends ActionType<GetWorkflowStateResponse> {
+    /** The name of this action */
+    public static final String NAME = TRANSPORT_ACTION_NAME_PREFIX + "workflow_state/get";
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/flowframework/transport/GetWorkflowStateAction.java:20]: nit: with different action names btw we might have a slight unintended consequence in terms of security if we also return the template body with the state later on. We might need to just add this in documentation but basically if a user has permission for workflow_state/get action but not `workflow/get` then they will still be able to get workflow with just the `workflow_state/get` action allowed, not too big of an issue I think but just calling it out

File: src/main/java/org/opensearch/flowframework/transport/GetWorkflowStateRequest.java [hunk 2]
```diff
@@ -17,9 +17,9 @@
 import java.io.IOException;
 
 /**
- * Transport Request to get a workflow or workflow status
+ * Transport Request to get a workflow status
  */
-public class GetWorkflowRequest extends ActionRequest {
+public class GetWorkflowStateRequest extends ActionRequest {
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/flowframework/transport/GetWorkflowStateRequest.java:22]: We now have `workflowRequest` and `GetWorkflowStateRequest` , do you think we should have a separate `GetWorkflowRequest` to differentiate get and post/put (create)?

Comment (joshpalis) [reply src/main/java/org/opensearch/flowframework/transport/GetWorkflowStateRequest.java:22]: Agreed, ideally we should separate out the `WorkflowRequest` class into `CreateWorkflowRequest` and `ProvisionWorkflowRequest` and have them used specifically for their corresponding transport actions. But that is out of the scope of this PR. which is only for the Get Workflow API

File: src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowAction.java [hunk 3]
```diff
@@ -55,6 +55,11 @@ public String getName() {
         return GET_WORKFLOW_ACTION;
     }
 
+    @Override
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowAction.java:58]: No problem moving this up but I'd like all our REST APIs to follow the same ordering

File: src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowAction.java [hunk 4]
```diff
@@ -68,17 +73,16 @@ protected BaseRestHandler.RestChannelConsumer prepareRequest(RestRequest request
 
             // Validate content
             if (request.hasContent()) {
-                throw new FlowFrameworkException("No request body present", RestStatus.BAD_REQUEST);
+                throw new FlowFrameworkException("Invalid request format", RestStatus.BAD_REQUEST);
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowAction.java:76]: Whatever we put here is ignored, see #274

**Review by dbwiddis — COMMENTED** [review]
Quick look but then I got lost in the file renames and changes.   I agree "Get" is poorly named for the status API but I'm not sure what all the new APIs do. Can you summarize quickly in the PR summary/first-comment?

File: src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowAction.java [hunk 5]
```diff
@@ -88,7 +92,7 @@ protected BaseRestHandler.RestChannelConsumer prepareRequest(RestRequest request
                     channel.sendResponse(new BytesRestResponse(ex.getRestStatus(), exceptionBuilder));
 
                 } catch (IOException e) {
-                    logger.error("Failed to send back provision workflow exception", e);
+                    logger.error("Failed to send back get template exception", e);
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowAction.java:95]: the GET api path looks like this: `_plugins/_flow_framework/workflow` , it would make more sense if we change this to "Failed to send back get workflow exception"? could be weird to do a GET operation on a workflow ID under `/workflow/` and see template in exception message

File: src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowStateAction.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+            if (!flowFrameworkFeatureEnabledSetting.isFlowFrameworkEnabled()) {
+                throw new FlowFrameworkException(
+                    "This API is disabled. To enable it, update the setting [" + FLOW_FRAMEWORK_ENABLED.getKey() + "] to true.",
+                    RestStatus.FORBIDDEN
+                );
+            }
+
+            // Validate content
+            if (request.hasContent()) {
+                throw new FlowFrameworkException("No request body present", RestStatus.BAD_REQUEST);
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowStateAction.java:71]: I assume same issue as @dbwiddis mentioned above https://github.com/opensearch-project/flow-framework/pull/273/files#r1423261483.

Can be fixed here or separately when fixing the issue.

File: src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowAction.java (hunk 5 above)
Comment (joshpalis) [reply src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowAction.java:95]: sure ill make this change

File: src/main/java/org/opensearch/flowframework/transport/GetWorkflowStateRequest.java (hunk 2 above)
Comment (amitgalitz) [reply src/main/java/org/opensearch/flowframework/transport/GetWorkflowStateRequest.java:22]: we also should have a getWorkRequest right eventually?

File: src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowAction.java (hunk 3 above)
Comment (joshpalis) [reply src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowAction.java:58]: I think the diff is wonky due to the file name changes. Essentially this new API `RestGetWorkflowTransportAction` is used to retrieve a template by workflow_id

File: src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowStateAction.java (hunk 6 above)
Comment (joshpalis) [reply src/main/java/org/opensearch/flowframework/rest/RestGetWorkflowStateAction.java:71]: @dbwiddis we can address this at a later time when actioning that issue

**Review by dbwiddis — APPROVED** [review]

**Review by amitgalitz — APPROVED** [review]
